### PR TITLE
MID-2810 Adding in some logic to ensure that creates happen before deletes

### PR DESCRIPTION
**Pull request by cjbooms**
This logic relies on the period sync being in place to ensure old Ingress resources are correctly deleted. Currently the sync period is 20 seconds, so all old Ingress resources will be deleted 20 second after creates have being processed.

Signed-off-by: Conor Gallagher <conor.gallagher@zalando.ie>

### CODE REVIEW

**Comment by cjbooms** [comment]
👎  Consider moving this logic into the `IngressDerivationChain` class, so that the new backend weights can be applied to the soon-to-be-deleted routes

File: src/it/scala/ie/zalando/fabric/gateway/TestUtils.scala [hunk 1]
```diff
@@ -43,9 +43,6 @@ object TestUtils {
     case object InvalidRequest extends ResourcePayload {
       def payload: String = Source.fromResource("synch/invalid.json").mkString
     }
-    case object Spp3240Payload extends ResourcePayload {
-      def payload: String = Source.fromResource("synch/spp3240OwnershipConfirmationDeserializationBug.json").mkString
-    }
```
Comment (cjbooms) [inline src/it/scala/ie/zalando/fabric/gateway/TestUtils.scala:48]: unrelated change. This wasn't used anywhere

Comment (irishshagua) [reply src/it/scala/ie/zalando/fabric/gateway/TestUtils.scala:48]: Maybe it was meant to be used in the  `SPP3240BugVerificationSpec.scala` test that was deleted?

File: src/main/scala/ie/zalando/fabric/gateway/service/ZeroDowntimeIngressTransitions.scala [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    ingressDerivationChain
+      .deriveRoutesFor(spec, metadata)
+      .map { desiredRoutes =>
+        val deleted = relativeComplement(desiredRoutes, existingRoutes)
+        val created = relativeComplement(existingRoutes, desiredRoutes)
+        if (created.isEmpty) desiredRoutes else desiredRoutes ++ deleted
+      }
+  }
+
+  private def relativeComplement(a: Seq[IngressDefinition], b: Seq[IngressDefinition]) =
```
Comment (irishshagua) [inline src/main/scala/ie/zalando/fabric/gateway/service/ZeroDowntimeIngressTransitions.scala:20]: maybe more descriptive names than a and b?

Comment (cjbooms) [reply src/main/scala/ie/zalando/fabric/gateway/service/ZeroDowntimeIngressTransitions.scala:20]: It's basically the exact set theory definition. 
Screenshot 2020-07-21 at 16 14 19

File: src/it/scala/ie/zalando/fabric/gateway/TestUtils.scala (hunk 1 above)
Comment (cjbooms) [reply src/it/scala/ie/zalando/fabric/gateway/TestUtils.scala:48]: It wasn't used in in there, that test was using the more common Gateway Spec which is tested repeatedly elsewhere

File: src/main/scala/ie/zalando/fabric/gateway/web/GatewayWebhookRoutes.scala [hunk 3]
```diff
@@ -19,24 +20,27 @@ trait GatewayWebhookRoutes extends JsonModels {
...
           trace("metacontroller_synchronization") { span =>
             entity(as[SynchRequest]) { synchRequest =>
               extractExecutionContext { ec =>
                 implicit val execCtxt: ExecutionContext = ec
-
                 complete {
-                  ingressDerivationLogic
-                    .deriveRoutesFor(synchRequest.controlledResource.spec, synchRequest.controlledResource.metadata)
-                    .map { routes =>
-                      val resp = SynchResponse(ingressDerivationLogic.deriveStateFrom(synchRequest.currentState), routes)
```
Comment (irishshagua) [inline src/main/scala/ie/zalando/fabric/gateway/web/GatewayWebhookRoutes.scala:35]: Maybe the deriveStateFrom function can be removed also?

File: src/it/scala/ie/zalando/fabric/gateway/TestUtils.scala (hunk 1 above)
Comment (irishshagua) [reply src/it/scala/ie/zalando/fabric/gateway/TestUtils.scala:48]: I meant, the names are the same, possibly it was a typo that it was using the happy path input

File: src/main/scala/ie/zalando/fabric/gateway/service/ZeroDowntimeIngressTransitions.scala (hunk 2 above)
Comment (irishshagua) [reply src/main/scala/ie/zalando/fabric/gateway/service/ZeroDowntimeIngressTransitions.scala:20]: eeep... so we're naming things mathematically now?

Comment (cjbooms) [reply src/main/scala/ie/zalando/fabric/gateway/service/ZeroDowntimeIngressTransitions.scala:20]: For pure set theory operations, absolutely I am. Scala already has `union`, `intersection` and `diff` defined, I can't imagine how ugly things would be if everyone invented their own terminology for Set Theory operations.

**Comment by cjbooms** [comment]
👍

**Comment by herojan** [comment]
👍

**Comment by irishshagua** [comment]
👍

**Comment by cjbooms** [comment]
👍
